src/coprocessor/codec/batch/column.rs

```diff
@@ -0,0 +1,875 @@
+        impl BatchColumn {
+            /// Extracts a slice containing the entire values of the column
+            /// in specific type.
+            ///
+            /// # Panics
+            ///
+            /// Panics if current column is does not match the type.
+            #[inline]
+            pub fn $name(&self) -> &[Option<$ty>] {
+                match self {
+                    BatchColumn::$ty(ref vec) => vec.as_slice(),
+                    other => panic!(
+                        "Cannot call `{}` over a {} column",
+                        stringify!($name),
+                        other.eval_type()
+                    ),
+                }
+            }
+
+            /// Extracts a mutable slice containing the entire values of the column
+            /// in specific type.
+            ///
+            /// # Panics
+            ///
+            /// Panics if current column is does not match the type.
+            #[inline]
+            pub fn $mut_name(&mut self) -> &mut [Option<$ty>] {
+                match self {
+                    BatchColumn::$ty(ref mut vec) => vec.as_mut_slice(),
+                    other => panic!(
+                        "Cannot call `{}` over a {} column",
+                        stringify!($name),
+                        other.eval_type()
+                    ),
+                }
+            }
+        }
+
+        impl AsRef<[Option<$ty>]> for BatchColumn {
+            #[inline]
+            fn as_ref(&self) -> &[Option<$ty>] {
+                self.$name()
+            }
+        }
+
+        impl AsMut<[Option<$ty>]> for BatchColumn {
+            #[inline]
+            fn as_mut(&mut self) -> &mut [Option<$ty>] {
+                self.$mut_name()
+            }
+        }
+    };
+}
+
+impl_as_slice! { Int, as_int_slice, as_mut_int_slice }
+impl_as_slice! { Real, as_real_slice, as_mut_real_slice }
+impl_as_slice! { Decimal, as_decimal_slice, as_mut_decimal_slice }
+impl_as_slice! { Bytes, as_bytes_slice, as_mut_bytes_slice }
+impl_as_slice! { DateTime, as_date_time_slice, as_mut_date_time_slice }
+impl_as_slice! { Duration, as_duration_slice, as_mut_duration_slice }
+impl_as_slice! { Json, as_json_slice, as_mut_json_slice }
+
+macro_rules! impl_push {
+    ($ty:tt, $name:ident) => {
+        impl BatchColumn {
+            /// Pushes a value into the column in specific type.
+            ///
+            /// # Panics
+            ///
+            /// Panics if current column does not match the type.
+            #[inline]
+            pub fn $name(&mut self, v: Option<$ty>) {
+                match self {
+                    BatchColumn::$ty(ref mut vec) => vec.push(v),
+                    other => panic!(
+                        "Cannot call `{}` over to a {} column",
+                        stringify!($name),
+                        other.eval_type()
+                    ),
+                };
+            }
+        }
+    };
+}
+
+impl_push! { Int, push_int }
```

Comment:
A small suggestion. Since you have already use macro, why not change to fit the format like `impl_push! { Real, real }`.

Reply:
It's not possible. Rust does not support concat identifiers.

Reply:
Macro `concat_idents` exists, but is not stable. Forget it.

Reply:
concat_idents won't work here, try it and you will know. There is a community crate which uses compiler plugin which can work, but have disadvantage, which is, requires the compiler plugin to be enabled at the site where *code is used*, which means, if it is put in a workspace, we requires tikv instead of the workspace to enable this plugin. This is really wired. In summary, currently there is no elegant solutions.

Reply:
I see.